syntax = "proto3";

package vector_service;

// Vector Service - Embedding generation and Knowledge Hub operations
service VectorService {
  // Embedding Generation
  rpc GenerateEmbedding(EmbeddingRequest) returns (EmbeddingResponse);
  rpc GenerateBatchEmbeddings(BatchEmbeddingRequest) returns (BatchEmbeddingResponse);
  
  // Tool Vector Operations (Roosevelt's Knowledge Hub Expansion!)
  rpc UpsertTools(UpsertToolsRequest) returns (UpsertToolsResponse);
  rpc SearchTools(SearchToolsRequest) returns (SearchToolsResponse);
  
  // Generic Vector Operations (Unified Qdrant Access)
  rpc UpsertVectors(UpsertVectorsRequest) returns (UpsertVectorsResponse);
  rpc SearchVectors(SearchVectorsRequest) returns (SearchVectorsResponse);
  rpc DeleteVectors(DeleteVectorsRequest) returns (DeleteVectorsResponse);
  rpc UpdateVectorMetadata(UpdateVectorMetadataRequest) returns (UpdateVectorMetadataResponse);
  
  // Collection Operations
  rpc CreateCollection(CreateCollectionRequest) returns (CreateCollectionResponse);
  rpc DeleteCollection(DeleteCollectionRequest) returns (DeleteCollectionResponse);
  rpc ListCollections(ListCollectionsRequest) returns (ListCollectionsResponse);
  rpc GetCollectionInfo(GetCollectionInfoRequest) returns (GetCollectionInfoResponse);
  
  // Cache Management
  rpc ClearEmbeddingCache(ClearCacheRequest) returns (ClearCacheResponse);
  rpc GetCacheStats(CacheStatsRequest) returns (CacheStatsResponse);
  
  // Health Check
  rpc HealthCheck(HealthCheckRequest) returns (HealthCheckResponse);
}

// ============================================================================
// Tool Vector Operations Messages
// ============================================================================

message ToolDefinition {
  string name = 1;
  string description = 2;
  string pack = 3;
  repeated string keywords = 4;
}

message UpsertToolsRequest {
  repeated ToolDefinition tools = 1;
}

message UpsertToolsResponse {
  bool success = 1;
  int32 count = 2;
  string error = 3;
}

message SearchToolsRequest {
  string query = 1;
  int32 limit = 2;
  float min_score = 3;
  string pack_filter = 4;
}

message ToolMatch {
  string name = 1;
  string description = 2;
  string pack = 3;
  float score = 4;
}

message SearchToolsResponse {
  repeated ToolMatch matches = 1;
  string error = 2;
}

// ============================================================================
// Embedding Generation Messages
// ============================================================================

message EmbeddingRequest {
  string text = 1;
  string model = 2;  // Optional, defaults to service config
}

message EmbeddingResponse {
  repeated float embedding = 1;
  int32 token_count = 2;
  string model = 3;
  bool from_cache = 4;  // Indicates if result came from cache
}

message BatchEmbeddingRequest {
  repeated string texts = 1;
  string model = 2;  // Optional
  int32 batch_size = 3;  // Optional, for parallel processing
}

message BatchEmbeddingResponse {
  repeated EmbeddingVector embeddings = 1;
  int32 total_tokens = 2;
  string model = 3;
  int32 cache_hits = 4;  // How many came from cache
  int32 cache_misses = 5;  // How many were generated
}

message EmbeddingVector {
  repeated float vector = 1;
  int32 index = 2;
  int32 token_count = 3;
  bool from_cache = 4;
}

// ============================================================================
// Cache Management Messages
// ============================================================================

message ClearCacheRequest {
  bool clear_embeddings = 1;  // Clear embedding cache
  string content_hash = 2;     // Optional: clear specific hash
}

message ClearCacheResponse {
  bool success = 1;
  int32 entries_cleared = 2;
  string error = 3;
}

message CacheStatsRequest {}

message CacheStatsResponse {
  int64 embedding_cache_size = 1;
  int64 embedding_cache_hits = 2;
  int64 embedding_cache_misses = 3;
  float cache_hit_rate = 4;
  int32 ttl_seconds = 5;
}

// ============================================================================
// Health Check Messages
// ============================================================================

message HealthCheckRequest {}

message HealthCheckResponse {
  string status = 1;  // "healthy", "degraded", "unhealthy"
  bool openai_available = 2;
  string service_version = 3;
  map<string, string> details = 4;
}

// ============================================================================
// Generic Vector Operations Messages
// ============================================================================

message VectorPoint {
  string id = 1;  // Point ID (UUID or hash)
  repeated float vector = 2;
  map<string, string> payload = 3;  // String-encoded metadata (JSON for complex types)
}

message UpsertVectorsRequest {
  string collection_name = 1;
  repeated VectorPoint points = 2;
}

message UpsertVectorsResponse {
  bool success = 1;
  int32 points_stored = 2;
  optional string error = 3;
}

message VectorFilter {
  string field = 1;
  string value = 2;
  string operator = 3;  // "equals", "contains", "in" (for arrays)
}

message SearchVectorsRequest {
  string collection_name = 1;
  repeated float query_vector = 2;
  int32 limit = 3;
  float score_threshold = 4;
  repeated VectorFilter filters = 5;
}

message VectorSearchResult {
  string id = 1;
  float score = 2;
  map<string, string> payload = 3;
}

message SearchVectorsResponse {
  bool success = 1;
  repeated VectorSearchResult results = 2;
  optional string error = 3;
}

message DeleteVectorsRequest {
  string collection_name = 1;
  repeated VectorFilter filters = 2;  // Delete by filter
}

message DeleteVectorsResponse {
  bool success = 1;
  int32 points_deleted = 2;
  optional string error = 3;
}

message UpdateVectorMetadataRequest {
  string collection_name = 1;
  repeated VectorFilter filters = 2;  // Which vectors to update
  map<string, string> metadata_updates = 3;  // New metadata values
}

message UpdateVectorMetadataResponse {
  bool success = 1;
  int32 points_updated = 2;
  optional string error = 3;
}

message CreateCollectionRequest {
  string collection_name = 1;
  int32 vector_size = 2;
  string distance = 3;  // "COSINE", "EUCLIDEAN", "DOT"
}

message CreateCollectionResponse {
  bool success = 1;
  optional string error = 2;
}

message DeleteCollectionRequest {
  string collection_name = 1;
}

message DeleteCollectionResponse {
  bool success = 1;
  optional string error = 2;
}

message ListCollectionsRequest {}

message CollectionInfo {
  string name = 1;
  int32 vector_size = 2;
  string distance = 3;
  int64 points_count = 4;
  string status = 5;
}

message ListCollectionsResponse {
  bool success = 1;
  repeated CollectionInfo collections = 2;
  optional string error = 3;
}

message GetCollectionInfoRequest {
  string collection_name = 1;
}

message GetCollectionInfoResponse {
  bool success = 1;
  optional CollectionInfo collection = 2;
  optional string error = 3;
}
